### PR TITLE
Consolidate website-tools into build-tools

### PR DESCRIPTION
**Pull request by sdake**
Now build-tools includes a ruby context, a nodejs context,
and includes various tools required to run the website build.

### CODE REVIEW

**Comment by geeknoid** [comment]
I'm experimenting with this to see if I can figure out the problem. One thing I noticed is we're not pinning all the versions. In particular, here are some versions we're using:

mdspell 1.3.1
mdl 0.5.0
htmlproofer 3.9.2
sass-lint 1.13.1
tslint 5.18.0

**Comment by geeknoid** [comment]
Previous image was around 600M, this pushes things to 1.19G or so

**Comment by sdake** [comment]
I think building nodejs from source would result in a smaller image.

This particular image size doesn't really matter, as we are caching the image in nearly al cases, right?

Cheers
-steve

**Comment by geeknoid** [comment]
I agree that size isn't everything...

**Comment by sdake** [comment]
The nodejs implementation takes 400mb.  The image looks like this without the nodejs context in the final image:

```
sdake/build-tools                                    1.0.0                                                              ae1dce568c8d        4 seconds ago       799MB
```

Here is the build tools image size history:
```
Stevens-MacBook-Pro:build-tools sdake$ docker images | grep build-tools
gcr.io/istio-testing/build-tools     2019-08-14          c0443ec59cf0        2 hours ago         1.19GB
gcr.io/istio-testing/build-tools     2019-08-13          512f7c6fbba0        30 hours ago        602MB
gcr.io/istio-testing/build-tools     2019-08-12          192928c7635d        2 days ago          602MB
gcr.io/istio-testing/build-tools     2019-08-09          64e550c41d6f        5 days ago          937MB
gcr.io/istio-testing/build-tools     2019-08-08          35d16310ac3b        6 days ago          628MB
```

Ruby consumes 200mb and nodejs consumes 400mb.  I think the nodejs implementation if compiled from source would be smaller - ruby certainly is.

Going to try.

File: docker/build-tools/Dockerfile [hunk 1]
```diff
@@ -96,16 +96,18 @@ FROM golang:1.12 as build_hugo_context
...
-RUN mv /tmp/hugo /out/go/bin
+RUN tar -xzvf /tmp/hugo_${HUGO_VERSION}_Linux-64bit.tar.gz --strip-components=1 -C /usr/local
 
 # Build a nodejs context
-FROM ubuntu:xenial as build_nodejs_context
-RUN apt-get update
-RUN apt-get -y install curl
-RUN curl -sL https://deb.nodesource.com/setup_10.x | bash -
-RUN apt-get -y install nodejs
+FROM ubuntu:xenial as node_tools_context
```
Comment (ericvn) [inline docker/build-tools/Dockerfile]: Update comment to be node tools instead of nodejs

File: docker/build-tools/Dockerfile [hunk 2]
```diff
@@ -96,16 +96,18 @@ FROM golang:1.12 as build_hugo_context
 ENV HUGO_VERSION=0.55.5
 RUN mkdir -p /out/go/bin
 ADD https://github.com/gohugoio/hugo/releases/download/v${HUGO_VERSION}/hugo_${HUGO_VERSION}_Linux-64bit.tar.gz /tmp
-RUN tar -xzvf /tmp/hugo_${HUGO_VERSION}_Linux-64bit.tar.gz -C /tmp
-RUN mv /tmp/hugo /out/go/bin
+RUN tar -xzvf /tmp/hugo_${HUGO_VERSION}_Linux-64bit.tar.gz --strip-components=1 -C /usr/local
```
Comment (ericvn) [inline docker/build-tools/Dockerfile]: Does the build_hugo_context need to use `golang:1.12`? We are also using `ubuntu:xenial`? Would that be a better fit. The steps could also be included in other contexts, but understand the reasoning behind separating out similar artifacts.

File: docker/build-tools/Dockerfile [hunk 3]
```diff
@@ -120,9 +170,23 @@ ENV GOPATH=/go
...
+
+# The build_nodejs_context needs attention - probably need to build nodejs from source
+#COPY --from=build_nodejs_context / /
+
+COPY --from=build_hugo_context /out/ /
 COPY --from=build_tools_context /out/ /
+COPY --from=ruby_tools_context /usr/local/bin /usr/local/bin
+COPY --from=ruby_tools_context /usr/local/lib /usr/local/lib
+COPY --from=node_tools_context /usr/local /usr/local
+COPY --from=node_tools_context /package.json /
```
Comment (geeknoid) [inline docker/build-tools/Dockerfile]: Both the ruby and the nodejs stuff will end up carrying a whole bunch of stuff redundantly within the image I think. Can we be more selective in the context creation and but just the new Ruby- and Node-related bits in an /out directory so we end up with no redundant state in layers?

Either that, or just create one more scratch context at the end and copy from the previously final context into it. That'll flatten everything.

Comment (geeknoid) [reply docker/build-tools/Dockerfile]: I just tried this and it actually made things bigger.

Shows how much I know about this...

File: docker/build-tools/Dockerfile (hunk 2 above)
Comment (sdake) [reply docker/build-tools/Dockerfile]: for hugo - we could use scratch - its jsut a build context.

File: docker/build-tools/Dockerfile [hunk 4]
```diff
@@ -90,11 +90,64 @@ RUN rm -fr /usr/local/go/api
...
+    typescript \
+    tslint \
+    markdown-spellcheck \
+    svgstore-cli \
+    svgo
+RUN npm install --save-dev @babel/core @babel/cli @babel/preset-env
+RUN npm install --save @babel/polyfill
+
+# TODO(sdake) Unclear if this is needed
+#RUN chmod -R 777 /root
```
Comment (geeknoid) [inline docker/build-tools/Dockerfile]: Delete?

File: docker/build-tools/Dockerfile [hunk 5]
```diff
@@ -118,11 +172,24 @@ ENV GO111MODULE=on
 ENV GOPROXY=https://proxy.golang.org
 ENV GOPATH=/go
 ENV GOCACHE=/gocache
-ENV PATH /go/bin:$PATH
+ENV PATH=/go/bin:$PATH
+ENV HOME=/root
```
Comment (geeknoid) [inline docker/build-tools/Dockerfile]: This isn't for Go's benefit, right? Maybe it should be in a separate section of env vars?

File: docker/build-tools/Dockerfile (hunk 4 above)
Comment (sdake) [reply docker/build-tools/Dockerfile]: deleted.

File: docker/build-tools/Dockerfile [hunk 6]
```diff
@@ -118,11 +172,24 @@ ENV GO111MODULE=on
...
+
+# Uhh ya. Ruby.
+ENV RUBYOPT="-KU -E utf-8:utf-8"
 
 COPY --from=base_os_context / /
 COPY --from=golang_context /usr/local/go/ /go/
+
+# The build_nodejs_context needs attention - probably need to build nodejs from source
+#COPY --from=build_nodejs_context / /
```
Comment (geeknoid) [inline docker/build-tools/Dockerfile]: Delete?

Comment (sdake) [reply docker/build-tools/Dockerfile]: deleted.

File: docker/build-tools/Dockerfile (hunk 5 above)
Comment (sdake) [reply docker/build-tools/Dockerfile]: ya - the golang context copies to /go/bin - so the PATH needs to be set appropriately.

File: docker/build-tools/Dockerfile (hunk 3 above)
Comment (sdake) [reply docker/build-tools/Dockerfile]: I've spent a fair amount of time optimizing the build contexts - to 950mb.  Pretty good sizing I think.  In _addition_ we were using some fellas PPA in the website tools for node.  Now we are using the upstream release from an upstream source (as well as ruby).

File: docker/build-tools/Dockerfile (hunk 2 above)
Comment (sdake) [reply docker/build-tools/Dockerfile]: agreed ubuntu:xenial would be a better fit - change made.

File: docker/build-tools/Dockerfile (hunk 1 above)
Comment (sdake) [reply docker/build-tools/Dockerfile]: addressed.

**Comment by geeknoid** [comment]
This image is now available as gcr.io/istio-testing/build-tools:2019-08-15

**Pull request merged**